cranelift/codegen/src/isa/unwind.rs

```diff
@@ -86,3 +89,89 @@ pub mod input {
+///
+/// Unwind pseudoinstructions are emitted inline by ABI code as it generates
+/// a prologue.
+///
+/// Thus, for the usual case, a prologue might look like (using x64 as
+/// an example):
+///
+/// ```plain
+/// push rbp
+/// unwind PushOldFP sp_offset_to_entry=8
+/// mov rbp, rsp
+/// unwind CreateFPFrame (BP=sp; [BP] = oldBP)
+/// sub rsp, 32
+/// mov [rsp+0], r12
+/// unwind SaveReg [BP-32], r12
+/// mov [rsp+8], r13
+/// unwind SaveReg [BP-24], r13
+/// ...
+/// ...
+/// mov rsp, rbp
+/// pop rbp
+/// unwind PopOldFP sp_offset_to_entry=0
+/// ret
+/// unwind Ret
+/// ```
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub enum UnwindInst {
+    /// The frame-pointer register for this architecture has just been
+    /// pushed to the stack. The FP has not been set to this frame yet.
+    /// The push decremented SP so that it is now `sp_offset_to_caller`
+    /// bytes below SP at the caller's callsite.
+    PushOldFP {
+        /// The offset from the current SP (after push) to the SP at
+        /// caller's callsite.
+        sp_offset_to_caller: u8,
+    },
+    /// The frame-pointer register for this architecture has just been popped
+    /// from the stack. The frame is now defined by SP again, and as above for
+    /// `PushOldFP`, `sp_offset_to_caller` gives the distance from current SP
+    /// (after pop) to SP at the caller's callsite.
+    PopOldFP {
+        /// The offset from the current SP (after pop) to the SP at
+        /// caller's callsite.
+        sp_offset_to_caller: u8,
+    },
+    /// The frame-pointer register for this architecture has just been
+    /// set to the current stack location, and the word at FP is the
+    /// previous FP value.
+    ///
+    /// The base of the frame is at `FP - fp_offset`. Note that only
+    /// positive offsets from the base of the frame are allowed for
+    /// `SaveReg`, but the FP register itself can point to any part
+    /// of the frame.
+    CreateFPFrame {
+        /// The offset from the start of frame to the FP value (and
+        /// current SP value).
+        fp_offset: u8,
+    },
+    /// The stack slot at the given offset from the frame base has
+    /// been used to save the given register.
+    ///
+    /// Given that `CreateFPFrame` has occurred first with some
+    /// `fp_offset`, `SaveReg` with `frame_offset` indicates that the
+    /// value of `reg` is saved on the stack at address `FP -
+    /// fp_offset + frame_offset`.
+    SaveReg {
+        /// The offset from the start of frame to this register's
+        /// stack location.
+        frame_offset: u8,
+        /// The saved register.
+        reg: RealReg,
+    },
+    /// A return has just occurred. This can be used by unwind
+    /// formats (like System V) to know where to delimit unwind-state
+    /// transitions around mid-function returns.
+    Ret,
```

Comment:
Since this and `PopOldFP` aren't actually used anywhere today, should they be removed for now?

I'd imagine that unwind info for the epilogue would also need to record that the saved registers no longer need to restored, but I don't think there's any emission for that just yet.

Reply:
Yep, this is left over from my intermediate PR state where I tried to handle epilogues too. Happy to remove!

Reply:
Yep, I've removed all traces of the epilogue attempt now.